### ISSUE TITLE
In generic class A<Ta>, method taking instance type B<Ta> (class B<Tb>) seems to accept B<any>.

### ISSUE THREAD

**Issue by dbarbeau** [Microsoft/TypeScript#13332]
Hello,

I hope the example is clearer than the title, I can't find a concise wording for the issue (I also hope this is not a user bug, I've been staring at this code for a while now).

**TypeScript Version:**  2.1.4

**Code**

```ts
class Inner<T>  {
	constructor() {
	}
}

class Outer<T> {
	add(_in: Inner<T>) {
	}

    push(_in: T) {
	}
}


let outer = new Outer< number >();

let innerBad = new Inner< string >();
let innerGood = new Inner< number >();

outer.add(innerBad); // should detect incompatible Inner types
outer.add(innerGood); // OK

outer.push(1); // OK
outer.push(""); // Error properly detected

```

**Expected behavior:**

Passing ``innerBad`` to ``outer.add`` should trigger a compilation error because of inconsistent types.

**Actual behavior:**

Typescript will not detect any error letting ``outer.add`` accept any ``Inner<XYZ>``.

**Comment by unional** [comment]
It's probably the compiler cannot decide what should `T` in `Inner<T>` be restricted to.

The following error out correctly:

```ts
class Inner<T>  {
  constructor(private x: T) {
  }
}

class Outer<T> {
  add(_in: Inner<T>) {
  }

  push(_in: T) {
  }
}


let outer = new Outer<number>();

let innerBad = new Inner<string>('1');
let innerGood = new Inner<number>(1);

outer.add(new Inner<string>('1')); // Error correctly
outer.add(innerGood); // OK

outer.push(1); // OK
outer.push("");
```

**Comment by dbarbeau** [comment]
Thanks! Indeed, it seems the compiler needs a little help. I'm not sure adding and argument of type T to the constructor is acceptable in my real code though. It works, but in my case it adds a useless parameter to Inner's constructor, which is unfortunate... However, this workaround will allow me to have my errors, but I look forward to removing that useless argument. ;)

**Comment by unional** [comment]
Well, the issue is not solved by "putting param in the constructor", it's solved because `Inner<T>` actually start using the type `T`.  Another example:

```ts
class Inner<T>  {
  multiple(x: T) {
    return x;
  }
}

class Outer<T> {
  add(_in: Inner<T>) {
  }
}


let outer = new Outer<number>();

let innerBad = new Inner<string>();
let innerGood = new Inner<number>();

outer.add(innerBad); // Error correctly
outer.add(innerGood); // OK

outer.push(1); // OK
outer.push("");
```

i.e., it should be working when you actually start implementing `Inner<T>`

**Comment by dbarbeau** [comment]
Yes, I understand, it's just that in my case putting the constructor was the most straight forward solution.

The real code is actually longer than that but Inner<T> acts like a type enforcing proxy-class, mostly passthrough operations. The important thing in it is the type parameter, not the values that the class might hold. In other words, it *is* implemented, but not in a way that allows TypeScript 2.1.4 to do the typechecking although type parameters are explicitely given.

Hum... while it did fix a simple case, I found more code which slips through. I'll come back tomorrow with a better code sample. 

Thanks again.

**Comment by RyanCavanaugh** [comment]
See https://github.com/Microsoft/TypeScript/wiki/FAQ#why-is-astring-assignable-to-anumber-for-interface-at-- . When you have a generic type that doesn't use its type parameter, the type parameter will not affect assignability relations.

**Comment by dbarbeau** [comment]
Hello,

I narrowed down the problem. In the real code, Inner does use its T (represented by `Inner.doSomething`). Type checks properly catch improper assignments when T is something like `number`, `string`, `boolean` or a class. However, the type checking misses improper assignments when T is a function signature, which is crucial to my code's type safetyness. 

```ts
class Inner<T>  {
  doSomething() : T|null {
    return null;
  }
}

class Outer<T> {
  add(_in: Inner<T>) {
  }
}


/* Working simple type example */
let outerBool = new Outer<boolean>();
let innerBoolBad = new Inner<string>();
let innerBoolGood = new Inner<boolean>();
outerBool.add(innerBoolBad); // OK: Error caught
outerBool.add(innerBoolGood); // OK

/* Working simple class example */
class A {
  something() {}
}

let outerClass = new Outer<A>();
let innerClassBad = new Inner<string>();
let innerClassGood = new Inner<A>();
outerClass.add(innerClassBad); // OK: Error caught
outerClass.add(innerClassGood); // OK

/* NOT working function example */
let outerFunc = new Outer<(i:number)=>void>();
let innerFuncBad = new Inner<()=>number>();
let innerFuncGood = new Inner<(i: number)=>void>();
outerFunc.add(innerFuncBad); // BAD: no error
outerFunc.add(innerFuncGood); // OK

```

**Comment by aluanhaddad** [comment]
Why should that be an error? The type signature states that it requires a function which can be passed a number and that it does not expect that function to return a value. `() => number` meets both of these requirements because it is valid in JavaScript to pass extra parameters to and ignore the values returned by functions.

**Comment by dbarbeau** [comment]
OK for not caring about return type, this matches what's done in other languages. And it matches my use too (which doesn't mean it's good though). However in Typescript, to specify the possibility to pass optional arguments wouldn't I have used `(i ?: number) => number` ? 

While, `Outer<(i ?: number)=>void>` would accept a `()=>void` signature, `Outer<(i : number)=>void>` is not expected to do so. 

Similarly, while `Outer<(...args : any[])=>void>` would accept both a `()=>void` signature or a `(a: number, b: string)=>void` signature, `Outer<()=>void>` is not expected to accept the latter.

In Typescript I mean. In javascript, sure, one can pass whatever to functions and hope for them to be ready for that. Using Typescript, I'm constraining the code to catch simple assignment errors. At least that's my primary use for TS' type system.

**Comment by aluanhaddad** [comment]
>However in Typescript, to specify the possibility to pass optional arguments wouldn't I have used (i ?: number) => number ?

That intuition makes sense but it's not actually correct in TypeScript. This is well documented but perhaps should be stated earlier on in the documentation
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

**Comment by unional** [comment]
Do you also mean this?

> Don’t ever have a generic type which doesn’t use its type parameter. See more details in TypeScript FAQ page.

**Comment by dbarbeau** [comment]
@aluanhaddad : Indeed, that doc is hidden in "Declaration Files" section, which is not what I'm writing, that's why I overlooked it. I think you're referring to 

> it’s always legal to provide a callback that accepts fewer arguments.

Which I'm trying right now to hammer deep into my head. It does kind of make sense. 
 ```ts
function anwserHelloToSomeone(msg: string) {
  return "Hello " + msg;
}

function answerHello(msg: string) {
  return "Hello";
}

function answer() {
  return "Hi";
}

let func : (msg: string) => string;
func = anwserHelloToSomeone;
func("George");

func = answerHello;
func("Brad");

func = answer;
func("Matt");
```

However, I'm still wondering. In my use case, passing shorter arity functions is highly indicative of an error. I need to enforce exact signature match to ensure the user is aware of what he's doing. I want the user to go for the `answerHello` way instead of the `answer` way in the example above. Can this be done?

@unional : That point is already addressed. My code uses its type parameter.

**Comment by dbarbeau** [comment]
Hello,

I'll close this report as the final question (enforce strict signature match using vanilla Typescript constructs) is now unrelated to the original issue.

Last two days I tried hacking my way through typescript to add such a feature - like I did for fun some time ago for mixins -  but got lost *. There are too many things to consider when hacking where I was hacking. Well, it was probably not a good idea anyway. 

For posterity, I was going for this type of uglyness :

```ts
let func1 !: (v1: number, v2: string) => boolean;
let func2 : (v1: number, v2: string) => boolean;
func1 = ()=>{return false} // error
func2 = ()=>{return false} // ok

function caller1(callback !: (v1: number, v2: string) => boolean) {/*...*/}
function caller2(callback : (v1: number, v2: string) => boolean) {/*...*/}
caller1(()=>{return false}) // error
caller2(()=>{return false}) // ok

```

Just changing `name : type` to `name !: type` (like an imperative "be strict!"), and restricted to plain function signature types to begin with. Might try again later though, the code is very clean, just huge.

\* 22053 lines in checker.ts ! tss tss tss ;)
